usr/sbin/p11sak/p11sak.c

```diff
@@ -939,50 +950,30 @@ static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_ULONG keylength,
     /* Boolean attributes (cannot be specified by user) */
     CK_BBOOL a_token = ckb_true; // always true
     CK_BBOOL a_private = ckb_true; // always true
-
-    /* Boolean attributes from user input */
-    CK_BBOOL a_modifiable = ckb_false;
-    CK_BBOOL a_derive = ckb_false;
-    CK_BBOOL a_sensitive = ckb_false;
-    CK_BBOOL a_encrypt = ckb_false;
-    CK_BBOOL a_decrypt = ckb_false;
-    CK_BBOOL a_sign = ckb_false;
-    CK_BBOOL a_verify = ckb_false;
-    CK_BBOOL a_wrap = ckb_false;
-    CK_BBOOL a_unwrap = ckb_false;
-    CK_BBOOL a_extractable = ckb_false;
     CK_ULONG bs = sizeof(CK_BBOOL);
 
     /* Non-boolean attributes */
     CK_ULONG a_value_len = keylength / 8;
 
-    CK_ATTRIBUTE tmplt[] = {
-            // boolean attrs
-            { CKA_TOKEN, &a_token, bs }, 
-            { CKA_PRIVATE, &a_private, bs }, 
-            { CKA_MODIFIABLE, &a_modifiable, bs }, 
-            { CKA_DERIVE, &a_derive, bs },
-            { CKA_SENSITIVE, &a_sensitive, bs }, 
-            { CKA_ENCRYPT, &a_encrypt, bs }, 
-            { CKA_DECRYPT, &a_decrypt, bs }, 
-            { CKA_SIGN, &a_sign, bs }, 
-            { CKA_VERIFY, &a_verify, bs }, 
-            { CKA_WRAP, &a_wrap, bs }, 
-            { CKA_UNWRAP, &a_unwrap, bs },
-            { CKA_EXTRACTABLE, &a_extractable, bs },
-            // non-boolean attrs
-            { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, 
-            { CKA_LABEL, label, strlen(label) } };
-    CK_ULONG num_attrs = sizeof(tmplt) / sizeof(CK_ATTRIBUTE);
-    CK_ULONG num_bools = num_attrs - 2;
-
-    /* set boolean attributes */
-    for (i = 0; i < (int) num_bools; i++) {
-        set_bool_attr_from_string(&tmplt[i], attr_string);
+    /* Standard template */
+    CK_ATTRIBUTE key_attr[4 + KEY_MAX_BOOL_ATTR_COUNT] = {
+        { CKA_TOKEN, &a_token, bs }, 
```

Comment:
Weh not just `sizeof(CK_ULONG)` instead of the `bs` variable?

Reply:
Done